[runtime] enhanced and refactored recently added Mutex abstraction

shawnce opened the pull request:
<!-- Please complete this template before creating pull request. -->
- added read / write lock support
- added non-fatal error support to allow use of mutex in fatal error reporting pathway
- isolated pthread implementation to it own header/cpp file pair
- expanded unit tests to cover new code as well as better test existing mutex
- removed a layer of complexity that added no real value
- added static allocation friendly versions of Mutex, ConditionVariable, and ReadWriteLock
#### Resolved bug number: ([SR-946](https://bugs.swift.org/browse/SR-946))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

shawnce commented:
Looking at conflict now.

shawnce commented on include/swift/Runtime/Mutex.h:
```diff
@@ -274,32 +288,227 @@ class Mutex {
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  void unlock() { Impl.unlock(); }
+
+  /// TODO
```
I will get to these header doc todo as soon as I can, I wanted to get this PR up to help out the effort in #1918.

gribozavr commented on include/swift/Runtime/Mutex.h:
```diff
@@ -10,124 +10,137 @@
 
-#if (defined(__APPLE__) || defined(__linux__) || defined(__CYGWIN__) || defined(__FreeBSD__))
-#define SWIFT_RUNTIME_MUTEX_HAVE_PHTREADS
+#if (defined(__APPLE__) || defined(__linux__) || defined(__CYGWIN__))
```
I'm afraid you erased the FreeBSD condition.

shawnce replied:
Yeah I just noticed that... need to get my preferred merge tool installed on my laptop (away from home again).

gribozavr commented on include/swift/Runtime/Mutex.h:
```diff
@@ -10,124 +10,137 @@
+
+/// A Condition that works with Mutex to allow – as an example – multi-threaded
+/// producers and consumers to signal each other in a safe way.
+class Condition {
```
Would it make more sense to call it `ConditionVariable`?  That seems like a more common name to me.

shawnce replied:
It may. I followed the pthread naming for now. I can look at the typical semantics for condition variable to ensure it doesn't deviate from the typical to much.

gribozavr commented:
Not a review, just triggering the CI.

@swift-ci Please test

shawnce commented:
Offline until morning (PST).

shawnce commented:
Looks like both Linux and Mac passed thanks for the early CI run.

jckarter commented on include/swift/Runtime/MutexPThread.h:
```diff
@@ -0,0 +1,102 @@
+
+private:
+  explicit MutexImpl(bool reportFatal, bool checked);
+  bool ReportFatal;
```
I don't think this flag needs to be changed at runtime. Can we make it a template parameter instead? Can it be defaulted?

shawnce replied:
We can.

shawnce replied:
Oh and I should note it is an internal constructor. The "wrapping" classes default things as makes sense.

jckarter replied:
Fair enough, but I still wouldn't want the unnecessary extra word of runtime state.

shawnce replied:
Will work on leveraging a template parameter for this and checked, unless you think I should just drop checked? It helps trap issues when running tests and in theory newly written code could leverage it to help flag issues in what they wrote before going into production, etc. The later could be predicated on build type as well. In really only applies to non-reclusive mutex since the other lock types need to maintain more state anyways giving you misuse detection "for free".

jckarter replied:
I'd be fine with dropping "checked", since we should still be able to test that we crash on misuse. What needs more state than `pthread_mutex_t` and `pthread_rwlock_t` already have? We shouldn't need a recursive mutex.

shawnce replied:
Sorry bad wording... rwlock already has to track sufficient context so it can catch typical misuse however non-recursive mutex purposely avoids tracking some state (since they doesn't need support recursive operation) which means doing something like m.lock(); m.lock() in the same thread (e.g. double lock) will result in a dead lock instead of an error noting it would deadlock. If you ask the mutex to 'PTHREAD_MUTEX_ERRORCHECK' then it will track additional state allowing it to return errors instead of some undefined, etc. behaviors. The check flag allows code to ask for the slightly more expensive state tracking to help flag misuse. I default to not engage the additional state tracking.

It is basically a debugging aid.

shawnce replied:
How important is the savings of moving the ReportFatal boolean to a non-type template parameter. In the runtime I see very few instances of mutex, etc. being created. Of course that could change if some aspect of the code for example decides to have a mutex per some swift object, etc... not sure how likely that is.

I ask because adding a template will cause some ripple effects in what I have implemented that I need to get some time to think about how best to manage. 

ReportError is only need for a specific pathway (not likely to change) so stepping back a bit I think I can avoid the template ripple and the boolean storage by providing a less generic way for this one pathway to avoid the fatalReport "looping" issue. I will try to get time a little later today to make this change, should have likely done it this "one of way" initially.

benlangmuir commented on unittests/runtime/Mutex.cpp:
```diff
@@ -13,12 +13,57 @@
+  std::vector<std::thread> threads;
+
+  // Block the threads we are about to create.
+  bool spinWait = true;
```
This should be an atomic bool.  Applies to several other tests.

shawnce replied:
The way the tests work we can be lazy about the atomic aspects of things. It will get the correct answer soon enough across all affected threads. I can change if folks feel that strongly.

I put a note a little higher up in the tests attempting to outline why I was using atomic operations.

benlangmuir replied:
Concurrently writing and accessing a non-atomic value is undefined behaviour.

shawnce replied:
In this situation only ever a single writer exists and all readers are looping over the value. You will either get the old value, the new value, or potentially a partial value (don't know if the last one is possible on any supported platforms for a simple bool, even in the partial value case things will function as expected). Also the optimizer should not be able to eliminate the loads so it will get the new value at some point. Is their some other issue that I should be educated on?

Anyway I have locally changed things to use atomics and now rerunning things, etc.

gribozavr replied:
It sounds exactly like a description of a read/write race to me.

> You will either get the old value, the new value, or potentially a partial value

You can't assume that.

jckarter replied:
It's *undefined*. The compiler optimizes assuming that a race on a nonatomic can't happen.

jckarter replied:
In particular, the reference to `spinWait` could legally be constant-folded at compile time to `true` in the lambda body.

shawnce replied:
> It sounds exactly like a description of a read/write race to me.

Except that the readers are looping so they may get the wrong old value at one put but then will get the correct value in the near future (sufficient for the usage pattern in this code), assuming the loads aren't optimized away, note barriers are happening in this code given what it is doing, I am running under that assumption in my use of lazy booleans. It could still fall in undefined territory despite my current understanding of the liberties of the optimizer in this particular code structure.

I have changed it to use std::atomic to avoid the potential pitfalls. Thanks for the information.

jckarter commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -349,42 +346,38 @@ TwoWordPair<const char *, uintptr_t>::Return
-  return Pair{result, size};
+
+  Pair pair;
+  TypeNameCacheLock.readWriteLock(
```
I think this function is abstracting a little too much control flow. I'd personally prefer separate `readLock` and `writeLock` blocks with explicit control flow in between. Is the intent to support a lock that can atomically promote a read lock to a write lock?

shawnce replied:
It a way it can promote for those simple situations it covers. It allows the "fast" path step that only needs the read lock to be repeated for you again with the write lock and only if the "fast" path step does work would it proceed to do the "mutation" work that required the write lock. It avoids having duplicate code.

shawnce replied:
@jckarter as a follow up do you want me to change this to a reader and writer block like before (repeating the reader block code)

jckarter replied:
Yeah, I didn't see any other uses of this function, so the abstraction doesn't seem worth it.

benlangmuir commented on unittests/runtime/Mutex.cpp:
```diff
@@ -362,3 +429,442 @@ TEST(MutexTest, ConditionLockOrWaitLockAndNotifyThreaded) {
+  std::vector<int> results;
+  results.assign(10, 0);
+
+  bool done = false;
```
Should be an atomic bool.  Applies to the other tests using "bool done" as well.

shawnce commented on include/swift/Runtime/MutexPThread.h:
```diff
@@ -0,0 +1,118 @@
+  UnsafeMutexImpl &operator=(UnsafeMutexImpl &&) = delete;
+
+public:
+  ~UnsafeMutexImpl() { pthread_mutex_destroy(&PThreadMutex); }
```
I am trying to find a definitive statement on calling 'pthread_mutex_destroy' on a pthread_mutex_t initialized using 'PTHREAD_MUTEX_INITIALIZER'. I see at least one statement say you don't need to (in context of being static mutex) but none that say you shouldn't. Anyone know what is expected?

jckarter replied:
If the intent here is for `UnsafeMutex` to be used for static initializations, then it should not have a destructor at all. Global destructors are evil. The constructor ought to be `constexpr` as well to ensure it doesn't install a global constructor or static `__cxa_guard`.

jckarter replied:
I think `StaticMutex` might be a better name too.

shawnce replied:
Yeah global destructors are evil in my experience, dealing with this by providing static versions as needed.

shawnce commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -349,42 +346,38 @@ TwoWordPair<const char *, uintptr_t>::Return
-  
-  static pthread_rwlock_t TypeNameCacheLock = PTHREAD_RWLOCK_INITIALIZER;
+
+  static ReadWriteLock TypeNameCacheLock;
```
I need a little guidance on code that is using a static Mutex or ReadWriteLock. Such code currently gets warning like the following. 
```
[8/126] Building CXX object stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Casting.cpp.o
/Users/shawnce/Development/apple_swift/swift/stdlib/public/runtime/Casting.cpp:350:24: warning: declaration requires an exit-time destructor [-Wexit-time-destructors]
  static ReadWriteLock TypeNameCacheLock;
```
So should I attempt to avoid static storage for locks, provide a variant that can be statically stored, suppress the warning, or? Note – while unlikely – non-pthread implementation may require or strong desire deallocation to clean up the lock implementations.

shawnce replied:
Decided to use `static Lazy<ReadWriteLock> TypeNameCacheLock;` for this situation. ..?

shawnce commented:
Think I dealt with most comments with an open question or two for folks. If possible getting a CI and benchmark run would be helpful.

jckarter commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -349,42 +346,39 @@ TwoWordPair<const char *, uintptr_t>::Return
-  
-  static pthread_rwlock_t TypeNameCacheLock = PTHREAD_RWLOCK_INITIALIZER;
+
+  static Lazy<ReadWriteLock> TypeNameCacheLock;
```
Can we have a `StaticReadWriteLock` too instead of making this lazy?

gribozavr commented:
@swift-ci Please test
@swift-ci Please benchmark

gribozavr commented:
@swift-ci Please test

jckarter commented:
In addition to CI, some should verify that this doesn't introduce any additional `-Wglobal-constructors` or `-destructors` warnings.

shawnce commented:
Looks like a lingering use of Condition instead of ConditionVariable not sure why my local benchmark run didn't flag it.

benlangmuir commented:
This is not a review of anything in this patch, but when writing tests for your synchronization primitives please bear in mind that 1) sleeping for XX ms is not sufficient to always have all your threads started and 2)  gtest ASSERT_* macros won't force the test to stop running if they happen inside a thread, since they work by using 'return'.

I just fixed a race in the previous tests with fa77a7a.

shawnce commented:
Yeah that test has been reworked in this patch for that very reason. I had made a comment on the prior PR about that issue after it was reported to be seen on Linux.

tinysun212 commented on include/swift/Runtime/Mutex.h:
```diff
@@ -186,121 +114,593 @@ class Mutex {
+  ///   either the read or write lock.
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  void unlock() { ReadWriteLockPlatformHelper::unlock(Handle); }
+
```
Could you separate this to two unlock's, for example ```readUnlock()```, ```writeUnlock()``` ?

I'm writing ```MutexWin32.cpp(.h)``` for class ```MutexImpl```. A problem I met is Win32 library (Slim Reader/Writer Locks) has two unlocks as ```<shared_mutex>```, ```ReleaseSRWLockExclusive()```  to ```shared_mutex::unlock()``` and ```ReleaseSRWLockShared()``` to ```shared_mutex::unlock_shared()```. It was tricky to select the proper function in one ```unlock()```. I think this trick is unnecessary if API is changed.

shawnce replied:
Yeah I can.

shawnce replied:
@tinysun212 glad you are looking into a Win32 version of this, sorry for changing things around some form what you likely first saw, decided to rework things to hopefully better support the variants that are potentially needed. Let me know if you hit any issues or have questions.

shawnce replied:
@tinysun212 I split unlock into unlockRead and unlockWrite. I am mostly now working on fleshing out the unit tests to cover both ReadWriteLock and StaticReadWriteLock. Let me know if you run across any other issues. I can also give you access to my repo if you want to drop something in directly.

tinysun212 commented on include/swift/Runtime/Mutex.h:
```diff
@@ -186,121 +114,593 @@ class Mutex {
+  void wait(ConditionVariable &condition) {
+    ConditionPlatformHelper::wait(condition.Handle, Handle);
+  }
 
```
Is it more natual that ```wait()``` method belongs to the condition variable class ?
```
pthread_cond_wait() in POSIX
condition_vairable::wait()  in <condition_variable>
WakeConditionVariable() in Win32
```

shawnce replied:
I had moved it over when I had wait also available on ScopedLock, etc. It makes more sense to move it back now.

tinysun212 commented:
I'm looking at ```Mutex::lockOrWait()```. Its body is equivalent to below code, is it right?

```
template <typename CriticalSection>
   void lockOrWait(ConditionVariable &condition,
                   CriticalSection criticalSection) {
     lock([&] {
       while (!criticalSection()) {
         wait(condition);
       }
     });
   }
```

```
  lock();
  while (!criticalSection()) {
    ConditionPlatformHelper::wait(condition.Handle, Handle);
  }
  unlock();
```

According to their manuals, all three condition variable's wait function (pthread, ```<condition_variable>```, Win32) release the associated lock. I think the second ```while loop``` will be running in unlocked.

shawnce commented:
> Upon successful return, the mutex shall have been locked and shall be owned by the calling thread.
I believe the above is universal for condition wait behavior (quoted from http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_cond_timedwait.html). You must hold the lock before calling wait, then before waiting (blocking calling thread) the wait function drops the lock, and before wait returns (assuming no error / timeout) it must reacquire the lock.

So while looping the lock is held except when inside of the wait call.

tinysun212 commented:
Oh I understand. I was wrong about pthread, ```<condition_variable>```, but Win32 ```SleepConditionVariableSRW()``` may differ.

shawnce commented:
I see the following community addition to the docs about SleepConditionVariableSRW(). I would be very surprised if it didn't acquire the lock before returning.

> Even after the wait time expires, this function does not return until it acquires the lock again no matter how long it takes. This thread must always release the lock regardless of the return value. The return value indicates that the conditional variable was set (true) or wait timed out (false + ERROR_TIMEOUT).

tinysun212 commented:
Yes, I confirmed with test code. It locks again !. MSDN document was poor. Thanks for reference.

tinysun212 commented:
I wrote ```MutexWin32.cpp(.h)``` and they passed your test code ```unittests/runtime/Mutex.cpp```. It was very simple. Thanks you.

shawnce committed:
[runtime] enhanced and refactored recently added Mutex abstraction

- added read / write lock support
- added non-fatal error support to allow use of mutex in fatal error reporting pathway
- isolated pthread implementation to it own header/cpp file pair
- expanded unit tests to cover new code as well as better test existing mutex
- removed a layer of complexity that added no real value

shawnce commented:
I think this is finally good to go. It will need a CI & benchmark before any attempt to merge.

I have held off removing ReadWriteLock::readWriteLock([]{}, []{}) for the moment since unlock got separated into readUnlock and writeUnlock in support of Win32 version of this code. The abstraction helps avoid mistakes with calling the wrong unlock (granted likely easily detected in testing) and of course avoids duplication of code in typical fast path / slow path type usage of read write locks. I will remove it folks want.

jckarter commented:
@swift-ci Please test
@swift-ci Please benchmark

jckarter commented:
@swift-ci Please test

jckarter commented:
@shawnce There's some linker heartburn:

```
Undefined symbols for architecture x86_64:
  "swift::Condition::notifyAll()", referenced from:
      __swift_allocBox_ in libswiftRuntime.a(HeapObject.cpp.o)
      _swift_getGenericMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getObjCClassMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getFunctionTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getTupleTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getExistentialMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      ...
  "swift::Condition::Condition()", referenced from:
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::BoxCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(HeapObject.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::ObjCClassCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::FunctionCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::TupleCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::MetatypeCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::GenericCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::WitnessTableCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      ...
  "swift::MutexImpl::lock()", referenced from:
      __swift_allocBox_ in libswiftRuntime.a(HeapObject.cpp.o)
      _swift_getGenericMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getObjCClassMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getFunctionTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getTupleTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getExistentialMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      ...
  "swift::MutexImpl::wait(swift::Condition&)", referenced from:
      __swift_allocBox_ in libswiftRuntime.a(HeapObject.cpp.o)
      _swift_getGenericMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getObjCClassMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getFunctionTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getTupleTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getExistentialMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      ...
  "swift::MutexImpl::unlock()", referenced from:
      __swift_allocBox_ in libswiftRuntime.a(HeapObject.cpp.o)
      _swift_getGenericMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getObjCClassMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getFunctionTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getTupleTypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      _swift_getExistentialMetatypeMetadata in libswiftRuntime.a(Metadata.cpp.o)
      ...
  "swift::MutexImpl::MutexImpl(bool)", referenced from:
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::BoxCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(HeapObject.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::ObjCClassCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::FunctionCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::TupleCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::MetatypeCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<(anonymous namespace)::ForeignTypeState>::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      swift::Lazy<swift::MetadataCache<(anonymous namespace)::GenericCacheEntry> >::defaultInitCallback(void*) in libswiftRuntime.a(Metadata.cpp.o)
      ...
ld: symbol(s) not found for architecture x86_64
```

shawnce commented:
@jckarter I spotted that I can't explain why that is happening. Can a plain CI run be done to confirm what I am seeing on my end (I don't see any linker issues, etc.). I will rerun benchmarks locally to confirm I see no issues (didn't see any last time).

jckarter commented:
@swift-ci Please test

shawnce commented:
@jckarter the issue looks specific to what the benchmark operation is doing. I can build --benchmark locally without issues. I suspect somehow a reference branch is being made from some invalid merge point that is resulting in a mix of old and new code trying to be compiled.

Any place I can look to see how a given CI build is setup, to understand the command executed, etc.?

jckarter commented:
@shahmishal Any idea why the benchmark runner would fail with linker errors when the normal CI works?

tinysun212 commented:
I looked the failed benchmark consol output https://ci.swift.org/job/swift-PR-osx-perf/35/console. I think the problem is in the ci build process.

The commit a0452be removed ```stdlib/public/runtime/Mutex.cpp``` and added ```stdlib/public/runtime/MutexPThread.cpp```, but the benchmark compiled both.
It seems it tested without ```MutexPThread.cpp```, and older source set. There is no ```MutexImpl``` class in current commit, isn't it?

```
[505/949] Building CXX object stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Mutex.cpp.o
.....
[30/167] Building CXX object stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/MutexPThread.cpp.o
.....
"swift::MutexImpl::lock()", referenced from:
```

shawnce commented:
@shahmishal @jckarter Do the build commands exist in git someplace for what the benchmark CI plan does? ..or does a lot of that live in Jenkins? Would a new pull request potentially side step some state that may have a hand in an incorrect merge being used?

shahmishal commented:
@swift-ci Please benchmark

shahmishal commented:
@shawnce Benchmark bot:

1. Checkout `master` branch  (Targeted Branch)
2. run `swift/utils/build-script -R --no-assertions --benchmark`
3. Checkout `origin/pr/1950/merge`
4. run `swift/utils/build-script -R --no-assertions --benchmark`
5. Upload the results to GitHub pull request.

jckarter commented:
@nadavrot Do these regressions look OK to you? It looks potentially legitimate, since both of the affected benchmarks are type-metadata-access-heavy.

nadavrot commented:
@jckarter I am seeing one improvement and one regression. The numbers we are seeing could be noise. In any case, we'll benchmark this change again with the internal setup of the performance benchmark. I think that this change can go in.

shawnce commented:
Also looking at the numbers it seems like some of the benchmarks result in low values that can cause what look like larger percentage changes. It seems like they should be enhanced to always achieve a threshold of say 100 if they are going to be measured in the way they are.

shawnce commented:
_bump_

jckarter commented:
Any remaining objections @gparker42 @nadavrot ?

jckarter commented:
I'm going to be out next week; @gparker42, if this looks good, do you mind merging?

tinysun212 commented:
I am building with your code on Cygwin. I found errors.

```
In file included from /cygdrive/c/Work/swift_msvc/swift/stdlib/public/runtime/Casting.cpp:25:
In file included from /cygdrive/c/Work/swift_msvc/swift/include/swift/Runtime/Mutex.h:24:
/cygdrive/c/Work/swift_msvc/swift/include/swift/Runtime/MutexPThread.h:49:32: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
  static constexpr MutexHandle staticInit() {
                               ^
/cygdrive/c/Work/swift_msvc/swift/include/swift/Runtime/MutexPThread.h:50:12: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
    return PTHREAD_MUTEX_INITIALIZER;
           ^
/usr/include/pthread.h:52:35: note: expanded from macro 'PTHREAD_MUTEX_INITIALIZER'
#define PTHREAD_MUTEX_INITIALIZER PTHREAD_NORMAL_MUTEX_INITIALIZER_NP
                                  ^
/usr/include/pthread.h:50:45: note: expanded from macro 'PTHREAD_NORMAL_MUTEX_INITIALIZER_NP'
#define PTHREAD_NORMAL_MUTEX_INITIALIZER_NP (pthread_mutex_t)19
                                            ^
```

shawnce commented:
@tinysun212 Any suggestions to how to correct? I don't currently have a usable Windows install with Cygwin (it will take me a bit to get it going). I guess Cygwin doesn't support staticly allocated pthread mutex?

tinysun212 commented:
I extracted some definitions from ```pthread.h``` in Cygwin.
I think you can reproduce above compile error with temporarily changing ```#include <pthread.h>``` to ```#include "swift/Runtime/_pthread.h"``` in ```Mutex.h/MutexPThread.h```.

```
typedef __uint32_t pthread_mutex_t;      /* identify a mutex */
typedef __uint32_t pthread_cond_t;       /* identify a condition variable */
typedef __uint32_t pthread_rwlock_t;         /* POSIX RWLock Object */

#define PTHREAD_COND_INITIALIZER (pthread_cond_t)21
#define PTHREAD_MUTEX_INITIALIZER PTHREAD_NORMAL_MUTEX_INITIALIZER_NP
#define PTHREAD_NORMAL_MUTEX_INITIALIZER_NP (pthread_mutex_t)19
#define PTHREAD_RWLOCK_INITIALIZER (pthread_rwlock_t)22

int pthread_cond_broadcast (pthread_cond_t *);
int pthread_cond_destroy (pthread_cond_t *);
int pthread_cond_init (pthread_cond_t *, const pthread_condattr_t *);
int pthread_cond_signal (pthread_cond_t *);
int pthread_cond_wait (pthread_cond_t *, pthread_mutex_t *);

int pthread_mutexattr_init (pthread_mutexattr_t *);
int pthread_mutexattr_settype (pthread_mutexattr_t *, int);
int pthread_mutexattr_destroy (pthread_mutexattr_t *);

int pthread_mutex_destroy (pthread_mutex_t *);
int pthread_mutex_init (pthread_mutex_t *, const pthread_mutexattr_t *);
int pthread_mutex_lock (pthread_mutex_t *);
int pthread_mutex_trylock (pthread_mutex_t *);
int pthread_mutex_unlock (pthread_mutex_t *);

int pthread_rwlock_destroy (pthread_rwlock_t *rwlock);
int pthread_rwlock_init (pthread_rwlock_t *rwlock, const pthread_rwlockattr_t *attr);
int pthread_rwlock_rdlock (pthread_rwlock_t *rwlock);
int pthread_rwlock_tryrdlock (pthread_rwlock_t *rwlock);
int pthread_rwlock_trywrlock (pthread_rwlock_t *rwlock);
int pthread_rwlock_unlock (pthread_rwlock_t *rwlock);
int pthread_rwlock_wrlock (pthread_rwlock_t *rwlock);
```

shawnce commented:
I guess we will have to predicate the use of constexpr on not Cygwin. It won't make cygwin any worse then it apparently was (doesn't have a "true" static allocated mutex, etc.) while allowing other platforms to correctly honor static allocation (under C++).

tinysun212 commented:
Sorry, I captured wrong header. In ```include/cygwin/types.h```, the ```typedef``` part is as follows,
```
typedef struct __pthread_condattr_t {char __dummy;} *pthread_condattr_t;
typedef struct __pthread_cond_t {char __dummy;} *pthread_cond_t;
typedef struct __pthread_rwlock_t {char __dummy;} *pthread_rwlock_t;
typedef struct __pthread_rwlockattr_t {char __dummy;} *pthread_rwlockattr_t;
```

I wrote simple source to test. 
```
$ cat tt.cpp

typedef struct __pthread_cond_t {char __dummy;} *pthread_cond_t;
#define PTHREAD_COND_INITIALIZER (pthread_cond_t)21

typedef pthread_cond_t ConditionHandle;

static constexpr ConditionHandle staticInit() {
  return PTHREAD_COND_INITIALIZER;
};

$ clang++ -c -std=c++11 tt.cpp
tt.cpp:7:34: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
static constexpr ConditionHandle staticInit() {
                                 ^
tt.cpp:8:10: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
  return PTHREAD_COND_INITIALIZER;
         ^
tt.cpp:3:34: note: expanded from macro 'PTHREAD_COND_INITIALIZER'
#define PTHREAD_COND_INITIALIZER (pthread_cond_t)21
                                 ^
1 error generated.
```

Do you have any idea to keep ```constexpr```?

shawnce committed:
[runtime] predicate the use of constexpr on what the platform can support

At the moment CYGWIN pthreads implementation doesn't support the use of constexpr for static allocation versions. The way they define things results in a reinterpret_cast which violates constexpr.

shawnce commented:
@tinysun212 I just pushed a change that should deal with the issue you reported under Cygwin on Windows. If possible can you try it out to verify it works as expected?

tinysun212 commented:
It is successfully compiled without any warning or error. Thanks for your patch.

shawnce commented:
@tinysun212 cool, you will need to update your Win32 implementation have the following in its header.

```
#define CONDITION_SUPPORTS_CONSTEXPR 1
#define MUTEX_SUPPORTS_CONSTEXPR 1
#define READWRITELOCK_SUPPORTS_CONSTEXPR 1
```

shawnce commented:
@gparker42 @jckarter ok I think this ready to go after a CI

gparker42 commented:
@swift-ci Please test
@swift-ci Please benchmark

gparker42 commented:
@swift-ci Please test

shawnce commented:
It looks like the Linux build issues continues to affect things?

`ERROR: testREPL (TestREPLArray.REPLArrayTestCase)`

gparker42 commented:
@swift-ci Please test

gparker42 commented on include/swift/Runtime/Mutex.h:
```diff
@@ -186,121 +114,687 @@ class Mutex {
+  /// returns `true` (done) readWriteLock returns to the caller. If the read
+  /// critical section returns `false` then the write critical section will
+  /// be called (with write lock still held).
+  ///
```
This function is too fancy and only used once as Joe noted below. Please delete it.

gparker42 commented on include/swift/Runtime/Mutex.h:
```diff
@@ -186,121 +114,687 @@ class Mutex {
-  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  /// Precondition: Mutex not held by this thread, undefined otherwise.
   template <typename CriticalSection>
   void lock(CriticalSection criticalSection) {
```
This naming convention should be changed to distinguish simple locking from lock-perform-unlock wrappers. Swift uses `with` for similar functions, which here would be `withLock(...)`, `withReadLock(...)`, etc.

gparker42 commented on include/swift/Runtime/Mutex.h:
```diff
@@ -186,121 +114,687 @@ class Mutex {
+
+  /// See ReadWriteLock::readWriteLock
+  template <typename ReadonlySection, typename WriteSection>
+  void readWriteLock(ReadonlySection readSection, WriteSection writeSection) {
```
Delete this one too.

gparker42 commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -165,25 +166,22 @@ struct crashreporter_annotations_t gCRAnnotations
-  pthread_mutex_lock(&crashlogLock);
-  
-  char *oldMessage = (char *)CRGetCrashLogMessage();
+  static StaticUnsafeMutex crashlogLock();
```
Add a comment here describing why this is an unsafe mutex.

gparker42 commented:
I noted a few stylistic things that should be changed, but this is good enough to accept now and polish later.

shawnce commented:
@gparker42 I will work on the changes outlined and open a new pull request when ready. Thanks
